### ISSUE TITLE
Dune should warn about missing dependencies that can be inferred

### ISSUE THREAD

**Issue by edwintorok** [ocaml/dune#9399]
## Desired Behavior

Would be nice if `dune` has shown a warning about the missing dependency even without having to turn sandboxing on explicitly in cases where it can infer that a source file or build target is referenced in a shell rule.

## Example

See https://github.com/ocaml-gospel/ortac/pull/187. In this case a source file was mentioned in a dune rule, but dune didn't include it as a dependency automatically:
```
(rule
 (mode promote)
 (alias runtest)
 (package ortac-examples)
 (deps
  (package ortac-qcheck-stm)
  varray_incl.ml)
 (targets varray_circular_tests.ml)
 (action
  (setenv
   ORTAC_ONLY_PLUGIN
   qcheck-stm
   (with-stdout-to
    %{targets}
    (run
     ortac
     qcheck-stm
     varray_circular_spec.mli
     "make 42 'a'"
     "char t"
     --include=varray_incl
     --quiet)))))
```

The only way to detect the missing dependency was to explicitly turn on sandboxing. The missing dependency was causing weird issues with caching (see linked PR for details) e.g:
*  I modified a file and 'dune runtest' worked even though it should've failed
* then did a 'dune clean && dune runtest --cache=disabled --force' which caused the test to fail
* then reverted everything back to the original source that was passing the tests
* `dune clean && dune runtest` was still failing even with `--force`, only another clean with `--cache=disabled` restored the correct state

This is quite an easy mistake to make, and I've seen the occasional weird issues with caching (always solved by fully specifying my dependencies), but didn't have quite a clear-cut issue on how it goes wrong until now.

**Comment by nojb** [comment]
> Would be nice if by default `dune` has shown a warning about the missing dependency in cases where it can infer that a source file or build target is referenced in a shell rule.

The issue is that Dune has no way of knowing that the referenced file is actually a dependency; it could be that the filename is simply recorded as a string somewhere, and that the referenced file is not actually needed...

**Comment by emillon** [comment]
I agree with @nojb, dune itself can not know this, otherwise it wouldn't be necessary to have `%{dep:...}` in the language.
I believe that this kind of check could be done in a separate tool that could look for common issues in `dune` files.
